rtos/Mutex.h

```diff
@@ -76,14 +76,21 @@ class Mutex : private mbed::NonCopyable<Mutex> {
      */
     osStatus unlock();
 
+    /** Check to see if the current thread is the owner of this mutex
+      @return  true if the mutex is owned by the current thread, false otherwise.
+     */
+    bool is_owner();
+
     ~Mutex();
 
 private:
     void constructor(const char *name = NULL);
+    friend class ConditionVariable;
 
     osMutexId_t               _id;
     osMutexAttr_t             _attr;
     mbed_rtos_storage_mutex_t _obj_mem;
+    uint32_t count;
```

Comment:
I would like to keep away from RTX specific access wherever possible.

Reply:
So would I, but in the case of mere asserts, I'm fairly happy to have a programming-error assert that is in some way platform-specific. I don't normally need my programming errors reported by every compilation target, so I'm not too worried if my deepest static asserts are only on the GCC build, for example.

So, yes, I dislike peeking RTX's structures, but I dislike adding a separate count variable to Mutex that is only going to be used for an assert even more.

How's about making the count facility conditional on NDEBUG/MBED_DEBUG or whatever it is that MBED_ASSERT is switched on? The need to query mutex count should only ever be needed for asserts. 

No recursive mutex I've ever seen makes it visible as API, as it would just encourage weird programming practices - making it a "diagnostic-only" API might be wise for that reason.

Reply:
By way of another reference point, I note that Thread has a few RTX-specific diagnostics like `used_stack`, which are `#ifdef MBED_OS_BACKEND_RTX5` flagged. But I think I prefer the idea of the portable-but-debug-only parallel count.

Reply:
I have strong feelings against both premature optimization and the use of defines to switching out code when not absolutely necessary. It isn't part of Mutex's public API so it shouldn't encourage weird programming practices. I would very much like the leave count as it is. Is this acceptable to you?

Reply:
I'd counter that this isn't optimisation, it's avoiding adding new run-time memory and CPU overhead to an quite important core primitive, which is only needed to produce an assert on a non-RTX build.

Good point that it's not public though, which does prevent casual abuse. Although that does kind of mean adding extra friends every time you want a similar assert? Not sure that's better than having it public with a DEBUG flag.

But sure, it can stay - I seem to be the only person objecting :)

Reply:
On the subject of unnecessary overhead and premature optimisation, I had a quick glance at Semaphore versus EventFlags to see if there might be a clear efficiency difference between the two. The only difference I note is that Semaphore::wait takes one more CMSIS-RTOS SVC calls, due to the odd count return.

But there's a clear memory optimisation issue across the whole API - Mutex, Semaphore and others all store their attributes in the object. That's totally unnecessary - it could just be a local variable in the constructor.

Could save 16 bytes per RTOS primitive there. Makes the word for the mutex count relatively insignificant.

Reply:
> Could save 16 bytes per RTOS primitive there. Makes the word for the mutex count relatively insignificant.

🍾